solutions/platinum/usaco-1332.mdx

```diff
@@ -0,0 +1,264 @@
+using namespace std;
+
+int main() {
+	string s;
+	cin >> s;
+	int n = s.size();
+	s = '$' + s;
+	long long ans = 0;
+	vector<long long> dp(6);
+	for (int i = 1; i <= n; i++) {
+		vector<long long> ndp(6);
+		// substring starting from index i
+		dp[0]++;
+		for (int j = 0; j < 6; j++) {
+			if (s[i] == "bessie"[j]) {
+				// Adding new character to our bessie sequence
+				ndp[(j + 1) % 6] += dp[j];
+				/*
+				 * if we reach the end of a bessie, we need to count the number
+				 * of substrings that contain it. dp[5] is the number of
+				 * potential start positions (n - i + 1) is the number of
+				 * potential end positions (any position including or after
+				 * index i). We can multiply those to get the number of
+				 * substrings that contain this bessie and add it to the answer.
+				 */
+				if (j == 5) { ans += (n - i + 1) * dp[5]; }
+			} else {
+				// The case if we can't add new character
+				ndp[j] += dp[j];
+			}
+		}
+		swap(ndp, dp);
+	}
+	cout << ans << endl;
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+Notice that we can represent the transition from $dp$ to $ndp$ as a system of linear reccurences which depend on the letter.
+
+B:$\\$
+$ndp[0] = 0 \\$
+$ndp[1] = dp[0] + 1 + dp[1] \\$
+$ndp[2] = dp[2] \\$
+$ndp[3] = dp[3] \\$
+$ndp[4] = dp[4] \\$
+$ndp[5] = dp[5] \\$
+$newans = ans \\$
+
+E:$\\$
+$ndp[0] = dp[0] + 1 + dp[5] \\$
+$ndp[1] = 0 \\$
+$ndp[2] = dp[1] + dp[2] \\$
+$ndp[3] = dp[3] \\$
+$ndp[4] = dp[4] \\$
+$ndp[5] = 0 \\$
+$newans = ans + (n - i + 1) * dp[5] \\$
+
+S:$\\$
+$ndp[0] = dp[0] + 1 \\$
+$ndp[1] = dp[1] \\$
+$ndp[2] = 0 \\$
+$ndp[3] = dp[2] \\$
+$ndp[4] = dp[3] + dp[4] \\$
+$ndp[5] = dp[5] \\$
+$newans = ans \\$
+
+I:$\\$
+$ndp[0] = dp[0] + 1 \\$
+$ndp[1] = dp[1] \\$
+$ndp[2] = dp[2] \\$
+$ndp[3] = dp[3] \\$
+$ndp[4] = 0 \\$
+$ndp[5] = dp[4] + dp[5] \\$
+$newans = ans \\$
+
+Identity (Any letter other than $b$, $e$, $s$, or $i$):$\\$
+$ndp[0] = dp[0] + 1 \\$
+$ndp[1] = dp[1] \\$
+$ndp[2] = dp[2] \\$
+$ndp[3] = dp[3] \\$
+$ndp[4] = dp[4] \\$
+$ndp[5] = dp[5] \\$
+$newans = ans \\$
+
+Representing these linear recurrences with matrices and multiplying all the matrices in the order from the one representing position $1$ to position $n$ would allow us to compute our answer. We can optimize updates by building a segment tree on the matrices and updating each node with a respective matrix representing a given letter.
+
+The $i$-th column $(1 \leq i \leq 6)$ of each matrix represents $dp[i - 1]$, the seventh column of each matrix represents the answer, and the eighth column of each matrix is a constant $1$ so we can add $1$ to $dp[0]$ at every index.
+
+Note that if the letter is $e$, we need to set $mat[5][6]$ as $n - i + 1$.
+
+To extract the answer from our matrices, we can multiply the $1 \times 6$ matrix representing the initial state of our dp $[0, 0, 0, 0, 0, 0, 1]$ by the product of all the matrices and look at the value in the $2$nd to last column, which is equivalent to the value of $mat[7][6]$.
```

Comment:
"and look at the value in the $2$nd to last column, which is equivalent to the value of $mat[7][6]$." this is kinda weird figure a different way to word or a different place to add it

Reply:
I reworded (and fixed part of the explanation).